source/common/http/filter_manager.h

```diff
@@ -304,6 +306,57 @@ class FilterManagerCallbacks {
    */
   virtual void encodeMetadata(MetadataMapVector& metadata) PURE;
 
+  /**
+   * Injects request trailers into a stream that originally did not have request trailers.
+   */
+  virtual void setRequestTrailers(RequestTrailerMapPtr&& request_trailers) PURE;
+
+  /**
+   * Passes ownership of received continue headers to the parent. This may be called multiple times
+   * in the case of multiple upstream calls.
+   */
+  virtual void setContinueHeaders(ResponseHeaderMapPtr&& response_headers) PURE;
+
+  /**
+   * Passes ownership of received response headers to the parent. This may be called multiple times
+   * in the case of multiple upstream calls.
+   */
+  virtual void setResponseHeaders(ResponseHeaderMapPtr&& response_headers) PURE;
+
+  /**
+   * Passes ownership of received response trailers to the parent. This may be called multiple times
+   * in the case of multiple upstream calls.
+   */
+  virtual void setResponseTrailers(ResponseTrailerMapPtr&& response_trailers) PURE;
+
+  // TODO(snowp): We should consider moving filter access to headers/trailers to happen via the
+  // callbacks instead of via the encode/decode callbacks on the filters.
+
+  /**
+   * The downstream request headers if set.
+   */
+  virtual RequestHeaderMapOptRef requestHeaders() PURE;
+
+  /**
+   * The downstream request trailers if present.
+   */
+  virtual RequestTrailerMapOptRef requestTrailers() PURE;
+
+  /**
+   * Retrieves a pointer to the continue headers set via setResponseHeaders.
+   */
+  virtual ResponseHeaderMapOptRef continueHeaders() PURE;
+
+  /**
+   * Retrieves a pointer to the response headers set via setResponseHeaders.
+   */
+  virtual ResponseHeaderMapOptRef responseHeaders() PURE;
```

Comment:
Consider a comment on lifetime - we've had at least one bug where a filter latched a reference to responseHeaders, response headers were replaced via sendLocalReply, and the stale reference caused a crash.
I'd also replace via setResponseHeaders -> via the most recent call to setResponseHeaders for clarity.